src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+                shift_equation_idx += 1
+
+        ell = 4 * shift_equation_idx
+        shift_eq[ell: ell + shift_equation_idx] = shift_b
+        shift_i[ell: ell + shift_equation_idx] = np.arange(shift_equation_idx)
+        shift_j[ell: ell + shift_equation_idx] = 2 * n_img
+        mask = np.where(shift_eq != 0)[0]
+        shift_eq = shift_eq[mask]
+        shift_i = shift_i[mask]
+        shift_j = shift_j[mask]
+        ell += shift_equation_idx
+        shift_equations = sps.csr_matrix((shift_eq, (shift_i, shift_j)),
+                                         shape=(shift_equation_idx, 2 * n_img + 1))
+
+        self.shift_equations = shift_equations
+        self.shift_equations_map = shift_equations_map
+
+    def estimate_shifts(self, memory_factor=10000):
         """
-        constructor of an object for estimating 3D orientations
+        Estimate 2D shifts in images using estimated rotations
+
+        The function computes the common lines from the estimated rotations,
+        and then, for each common line, estimates the 1D shift between its two
+        Fourier rays (one in image i and one in image j). Using the common
+        lines and the 1D shifts, the function solves the least-squares
+        equations for the 2D shifts.
+        This function processes the (Fourier transformed) images exactly as the
+        `build_clmatrix` function.
+
+        :param memory_factor: If there are N projections, then the system of
+            equations solved for the shifts is of size 2N x N(N-1)/2 (2N
+            unknowns and N(N-1)/2 equations). This may be too big if N is
+            large. If `memory_factor` between 0 and 1, then it is the
+            fraction of equation to retain. That is, the system of
+            equations solved will be of size 2N x N*(N-1)/2*`memory_factor`.
+            If `memory_factor` is larger than 100, then the number of
+            equations is estimated in such a way that the memory used by the equations
+            is roughly `memory_factor megabytes`. Default is 10000 (use all equations).
+            The code will generate an error if `memory_factor` is between 1 and 100.
+        :return: Estimated shifts for all images
         """
-        pass
 
-    def get_commonlines(self):
+        if memory_factor < 0 or (memory_factor > 1 and memory_factor < 100):
+            logger.error('Subsampling factor must be between 0 and 1 or larger than 100.')
+
+        n_theta = self.n_theta // 2
+        n_img = self.n_img
+        rotations = self.rotations
+
+        pf = self.pf.copy()
+
+        # Estimate number of equations that will be used to calculate the shifts
+        n_equations = self._estimate_num_shift_equations(n_img, memory_factor)
+        # Allocate local variables for estimating 2D shifts based on the estimated number
+        # of equations. The shift equations are represented using a sparse matrix,
+        # since each row in the system contains four non-zeros (as it involves
```

Comment:
There are only four unknown, the last `n_equations` elements of `shift_eq,` `shift_i`, and  `shift_j` are used to store the information of  the right side of sytem Ax = b. But we also define `shift_b` so the information is redundant. I will remove it in new version.

Reply:
Actually, since the `estimate_shifts` function return the `shift_equations` object, it is better to contain the `b` information too. I will keep the original form.

Reply:
Ok. Might be better note this in the comment and write it as `4 * n_equations + n_equations` then. Honestly, I find this baking of the right-hand side into the left-hand side confusing. Is this a requirement to make the linear solve work? If not, can we specify the right-hand side separately?

Reply:
The information on the right hand side of `shift_b` is not necessary if we keep additional copy of `shift_b`. This information was removed from `shift_equations` in a recent commit.